File: ShortestWordDistanceII.py
from collections import defaultdict

class WordDistance(object):
    def __init__(self, words):
        """
        initialize your data structure here.
        :type words: List[str]
        """
        self.l = len(words)
        self.where = defaultdict(list)
        for i in range(len(words)):
            w = words[i]
            self.where[w].append(i)
        

    def shortest(self, word1, word2):
        """
        Adds a word into the data structure.
        :type word1: str
        :type word2: str
        :rtype: int
        """
        l1, l2 = self.where[word1], self.where[word2]
        min_dist = self.l
        max_ofst = -1
        i, j = 0, 0
        while i<len(l1) and j<len(l2):
            min_dist = min(min_dist, abs(l1[i]-l2[j]))
            if l1[i]>l2[j]:
                j += 1
            else:
                i += 1
        return min_dist
            
        
        


# Your WordDistance object will be instantiated and called as such:
# wordDistance = WordDistance(words)
# wordDistance.shortest("word1", "word2")
# wordDistance.shortest("anotherWord1", "anotherWord2")

File: shortestWordDistanceII.kt
class WordDistance(wordsDict: Array<String>) {
    val map = hashMapOf<String, List<Int>>()
    init{
        for(i in 0 until wordsDict.size){
            val list = map.getOrDefault(wordsDict[i], listOf()).toMutableList()
            list.add(i)
            map[wordsDict[i]] = list.toList()
        }
    }

    fun shortest(word1: String, word2: String): Int {
        val list1 = map[word1]?: listOf()
        val list2 = map[word2]?: listOf()
        return findClosest(list1, list2)
    }
    private fun findClosest(list1: List<Int>, list2: List<Int>): Int{
        var result = Int.MAX_VALUE
        var i = 0; var j = 0
        while(i < list1.size && j < list2.size){
            result = min(result, abs(list1[i] - list2[j]))
            if(list1[i] > list2[j]) j++
            else i++
        }
        return result
    }

}

/**
 * Your WordDistance object will be instantiated and called as such:
 * var obj = WordDistance(wordsDict)
 * var param_1 = obj.shortest(word1,word2)
 */
